staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -483,42 +501,71 @@ impl VhostUserCanBackend {
         }
 
         for desc_chain in requests {
-            let descriptors: Vec<_> = desc_chain.clone().collect();
-
-            if descriptors.len() != 2 {
-                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
-            }
+            let atomic_mem = self.mem.as_ref().unwrap().memory();
 
-            let desc_request = descriptors[0];
-            if desc_request.is_write_only() {
+            let Ok(mut reader) = desc_chain.clone().reader(&atomic_mem) else {
+                trace!("Error::UnexpectedWriteOnlyDescriptor");
                 return Err(Error::UnexpectedWriteOnlyDescriptor(0));
-            }
+            };
 
-            if (desc_request.len() as usize) > size_of::<VirtioCanFrame>()
-                || (desc_request.len() as usize) < (size_of::<VirtioCanFrame>() - 64)
-            {
+            let Ok(mut writer) = desc_chain.clone().writer(&atomic_mem) else {
+                trace!("Error::UnexpectedReadableDescriptor");
+                return Err(Error::UnexpectedReadableDescriptor(0));
+            };
+
+            let reader_len = reader.available_bytes();
+            if reader_len < size_of::<VirtioCanHeader>() {
                 trace!(
                     "desc_len: {}, virtio_can_len: {:?}\n",
-                    desc_request.len() as usize,
-                    size_of::<VirtioCanFrame>()
+                    reader_len,
+                    size_of::<VirtioCanHeader>()
+                );
+                return Err(Error::UnexpectedRxDescriptorSize(
+                    size_of::<VirtioCanHeader>(),
+                    reader_len as u32,
+                ));
+            }
+
+            let writer_len = writer.available_bytes();
```

Comment:
I think you could do the same here, instead of checking the available bytes, you can just fails qhen you try to write later.